src/helper/remote_config/asm_data_listener.cpp

```diff
@@ -0,0 +1,142 @@
+                    expiration.value()->value.GetInt();
+            }
+        } else {
+            rule_data.data.insert({value.value()->value.GetString(),
+                {value.value()->value.GetString(),
+                    expiration.value()->value.GetInt()}});
+        }
+    }
+}
+
+dds::parameter rules_to_parameter(
+    const std::unordered_map<std::string, rule_data> &rules)
+{
+    dds::parameter parameter = dds::parameter::map();
+    dds::parameter rules_data = dds::parameter::array();
+    for (const auto &[key, rule] : rules) {
+        dds::parameter parameter_rule = dds::parameter::map();
+
+        parameter_rule.add("id", dds::parameter::string(rule.id));
+        parameter_rule.add("type", dds::parameter::string(rule.type));
+
+        // Data
+        dds::parameter data = dds::parameter::array();
+        for (const auto &[value, data_entry] : rule.data) {
+            dds::parameter data_parameter = dds::parameter::map();
+            data_parameter.add("expiration",
+                dds::parameter::uint64(data_entry.expiration, false));
+            data_parameter.add("value", dds::parameter::string(value));
+            data.add(std::move(data_parameter));
+        }
+        parameter_rule.add("data", std::move(data));
+
+        rules_data.add(std::move(parameter_rule));
+    }
+
+    parameter.add("rules_data", std::move(rules_data));
+
+    return parameter;
+}
+
+void dds::remote_config::asm_data_listener::on_update(const config &config)
+{
+    std::unordered_map<std::string, rule_data> rules_data = {};
+    rapidjson::Document serialized_doc;
+    if (!json_helper::get_json_base64_encoded_content(
+            config.contents, serialized_doc)) {
+        throw error_applying_config("Invalid config contents");
+    }
+
+    auto rules_data_itr = json_helper::get_field_of_type(
+        serialized_doc, "rules_data", rapidjson::kArrayType);
+    if (!rules_data_itr) {
+        throw error_applying_config("Invalid config json contents: "
+                                    "rules_data key missing or invalid");
+    }
+
+    for (rapidjson::Value::ConstValueIterator itr =
+             rules_data_itr.value()->value.Begin();
+         itr != rules_data_itr.value()->value.End(); ++itr) {
+        if (!itr->IsObject()) {
+            throw error_applying_config("Invalid config json contents: "
+                                        "rules_data entry invalid");
+        }
+
+        auto id_itr =
+            json_helper::get_field_of_type(itr, "id", rapidjson::kStringType);
+        auto type_itr =
+            json_helper::get_field_of_type(itr, "type", rapidjson::kStringType);
+        auto data_itr =
+            json_helper::get_field_of_type(itr, "data", rapidjson::kArrayType);
+        if (!id_itr || !type_itr || !data_itr) {
+            throw error_applying_config(
+                "Invalid config json contents: "
+                "rules_data missing a field or field is invalid");
+        }
+
```

Comment:
You can always assign `std::optional->value() `to a reference if you want the code to look a bit neater.

Reply:
Can you expand on this plz?

Reply:
Done